### PR TITLE
[INFINITY-1406] Change modify_master to act as a context manager, to get write-on-descope

### PR DESCRIPTION
**Pull request by jsrodman**
@benclarkwood I haven't yet done the bugfixing round, but I'm showing this now so you can critique if desired.

### CODE REVIEW

**Comment by benclarkwood** [comment]
Am cool with this approach.

**Comment by benclarkwood** [comment]
I am fine with how this works. Deferring to @mrbrowning for any Python quibbles.

File: tools/modify_master.py [hunk 1]
```diff
@@ -13,98 +28,171 @@
...
+            raise RuntimeError("Unable to get current envvars")
+
+        logger.info("Current envvars:\n{}".format(out))
+        envvars, commented_lines = self._decode_master_config(out)
+        self.master_envvars = envvars
+        self.commented_lines = commented_lines
+
+    def _modify_envvars(self, modifier_f):
+        """Performs a transformation on the mesos master environment
+        variables.  Accepts a funciton which accepts the mesos environment
```
Comment (mrbrowning) [inline tools/modify_master.py]: Typo, `funciton`

File: tools/modify_master.py [hunk 2]
```diff
@@ -13,98 +28,171 @@
...
+        before the transformation, and returns the new environment.  The
+        function should accept and return dicts of str:str
+
+        :param modifier_f: The function to apply the transformation to the envvar state
+        :type modifier_f: func
+        """
```
Comment (mrbrowning) [inline tools/modify_master.py]: What's the value of this pseudo-immutable treatment of the `master_envvars` member? Couldn't we just modify the dict in place?

**Comment by jsrodman** [comment]
Currently the intent was to change some settings on the master, not to replace the settings entirely.  If we wanted to wipe what was there, accepting a default dict might work, but I'd probably want the consumer to explicitly document that the prior state is being wiped with an explicit action.

File: tools/modify_master.py (hunk 2 above)
Comment (jsrodman) [inline tools/modify_master.py]: We could modify the dict in place, but it felt gross to me to have the functor dustructively modifying an object it doesn't really know the lifetime of.
I could move the protection against this to _modify_envvars(), or i could remove, but this felt more right to me than a destructive change that could break on refactor in weird ways.

**Pull request merged**